duplicate documents getting generated on save

Hi there 👋 Thanks for the amazing library - it saves my day! 🥳 

I'm having this weird issue I can't understand how to debug properly. I am using Next.js with the appropriate setup. I have a collection and it works like a charm when I start the next app for the first time. I see my collection of files and they render on the front page properly.

But then I go and just _save_ one of my files again. Regardless if I make changes or not. Then I have my file duplicated inside the collection. Here's the output (which is also weird):

```bash
# here I've hit ctrl+s inside my content-collections.ts
... config changed
build started ...
... finished build of 1 collection and 9 documents in 6ms
# all good so far

# here I hit ctrl+s inside file1.json, without making any changes
... file src\content\file1.json changed
build started ...
... finished build of 1 collection and 10 documents in 5ms
# somehow documents got 10, even though I haven't created another file

# weirdly enough, if I hit ctrl+s again (second time), two events get fired:
... file src\content\file1.json changed
build started ...
... finished build of 1 collection and 10 documents in 5ms
... file src\content\file1.json changed
build started ...
... finished build of 1 collection and 10 documents in 5ms

# hitting ctrl+s more times results in just two events getting fired..

# but if I ctrl+save my content-collections.ts again, I see this:
... config changed
build started ...
... finished build of 1 collection and 9 documents in 7ms
```

I'm seeing this on Windows 11, TS 5.6.3, with this following project setup:
```json
"devDependencies": {
  "@content-collections/core": "^0.7.2",
  "@content-collections/next": "^0.2.3",
  "@types/node": "^20",
  "@types/react": "^18",
  "@types/react-dom": "^18",
  "eslint": "^8",
  "eslint-config-next": "15.0.1",
  "postcss": "^8",
  "tailwindcss": "^3.4.1",
  "typescript": "^5"
}
```

Any ideas on how to troubleshoot it further? I'll probably clone the repo tomorrow and play with it, but I'd appreciate any hints upfront 😉

That's what I thought. Unfortunately I couldn't dig much deeper. I cloned the project and tried running the tests. The majority of them are failing, mostly because of this:

```
X [ERROR] The entry point "C:\\content-collections\\packages\\core\\src\\__tests__\\config.001.ts" cannot be marked as external
```

It leads to `esbuild.ts::56` where it's calling the `build` function. The path there is correct: `C:\content-collections\packages\core\src\__tests__\config.001.ts`, but it seems it cannot make it work. Running the `dev` tasks fails with the same reason 😞 

---

Apart from that I am able to build the project and use some sample setup like creating a new collection and printing its internals. Not of much help though.

The only thing that comes to my mind is create a new empty project, building the source code, copying it inside the project and initializing it using the import from the new `dist` I've created. Then trying the watcher.
Any ideas on how to test/debug it better?

I can also confirm tests are running okay now, except the `writer :: os specific` ones:

```
 FAIL  src/writer.test.ts > writer > os specific > should write import with / instead of \ on windows
AssertionError: expected 'import configuration from "../C:\User…' to contain 'import configuration from "./sub/conf…'

- Expected
+ Received

- import configuration from "./sub/config.ts";
+ import configuration from "../C:\Users\cnn\AppData\Local\Temp\vitest-XXXXXXadCcMt\sub\config.ts";
+ import { GetTypeByName } from "@content-collections/core";
+
+ export type Test = GetTypeByName<typeof configuration, "test">;
+ export declare const allTests: Array<Test>;
+
+ export {};
+
```

--- a/.gitattributes
+++ b/.gitattributes
@@ -1,0 +1,3 @@
+# use same eol as editorconfig and prettier
+* text eol=lf
+

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -17,6 +17,7 @@ jobs:
     env:
       TURBO_TOKEN: ${{ secrets.TURBO_TOKEN }}
       TURBO_TEAM: ${{ vars.TURBO_TEAM }}
+      CC_OS: ${{ matrix.os }}
 
     steps:
       - name: Check out code

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -10,7 +10,10 @@ jobs:
   build:
     name: Build and Test
     timeout-minutes: 15
-    runs-on: ubuntu-latest
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        os: [ubuntu-latest, windows-latest]
     env:
       TURBO_TOKEN: ${{ secrets.TURBO_TOKEN }}
       TURBO_TEAM: ${{ vars.TURBO_TEAM }}

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -12,6 +12,7 @@ jobs:
     timeout-minutes: 15
     runs-on: ${{ matrix.os }}
     strategy:
+      fail-fast: false
       matrix:
         os: [ubuntu-latest, windows-latest]
     env:

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -49,7 +49,7 @@ jobs:
         run: pnpm build
 
       - name: Install Playwright Browsers
-        run: pnpm playwright install --with-deps
+        run: pnpm playwright install chromium --with-deps
 
       - name: Test
         run: pnpm test

--- a/packages/core/src/collector.test.ts
+++ b/packages/core/src/collector.test.ts
@@ -1,3 +1,4 @@
+import path from "path";
 import { beforeEach, describe, expect, it } from "vitest";
 import { createCollector } from "./collector";
 import { Events, createEmitter } from "./events";
@@ -213,6 +214,24 @@ describe("collector", () => {
       ]);
 
       expect(collection?.files).toHaveLength(0);
+    });
+
+    it("should use native separator for collected paths", async () => {
+      // tinyglobby uses path.posix internally, so we have to convert the paths
+      // this test would fail on Windows if we didn't convert the paths
+      const [collection] = await collect([
+        {
+          directory: "./__tests__/sources/",
+          include: "test/*.md",
+          parser: "frontmatter",
+        },
+      ]);
+
+      const paths = collection?.files.map((file) => file.path);
+      expect(paths).toEqual([
+        path.join("test", "001.md"),
+        path.join("test", "002.md"),
+      ]);
     });
   });
 

--- a/packages/core/src/collector.ts
+++ b/packages/core/src/collector.ts
@@ -1,10 +1,10 @@
 import { readFile } from "fs/promises";
-import path from "path";
+import path from "node:path";
 import { glob } from "tinyglobby";
 import { Emitter } from "./events";
 import { parsers } from "./parser";
 import { CollectionFile, FileCollection } from "./types";
-import { isDefined, orderByPath } from "./utils";
+import { isDefined, orderByPath, posixToNativePath } from "./utils";
 
 export type CollectorEvents = {
   "collector:read-error": {
@@ -98,7 +98,7 @@ export function createCollector(emitter: Emitter, baseDirectory: string = ".") {
       ignore: createIgnorePattern(collection),
     });
     const promises = filePaths.map((filePath) =>
-      collectFile(collection, filePath),
+      collectFile(collection, posixToNativePath(filePath)),
     );
 
     const files = await Promise.all(promises);

--- a/packages/core/src/configurationReader.test.ts
+++ b/packages/core/src/configurationReader.test.ts
@@ -92,13 +92,13 @@ describe("configurationReader", () => {
 
   it("should throw an error if the config file does not exists", async () => {
     await expect(config("non-existing")).rejects.toThrowError(
-      /configuration file .*\/non-existing does not exist/,
+      /configuration file .*non-existing does not exist/,
     );
   });
 
   it("should throw an error if the config file is invalid", async () => {
     await expect(config("invalid")).rejects.toThrowError(
-      /configuration file .*\/invalid is invalid/,
+      /configuration file .*invalid is invalid/,
     );
   });
 });

--- a/packages/core/src/esbuild.ts
+++ b/packages/core/src/esbuild.ts
@@ -2,7 +2,7 @@ import { loadTsConfig, match, tsconfigPathsToRegExp } from "bundle-require";
 import { Plugin, build } from "esbuild";
 import { dirname, join } from "node:path";
 
-// the code to handle tsconfig paths is inspired by the awesome tsup project
+// the code to handle externals is mostly the one which is used by the awesome tsup project
 // https://github.com/egoist/tsup
 
 function tsconfigResolvePaths(configPath: string) {
@@ -13,9 +13,7 @@ function tsconfigResolvePaths(configPath: string) {
   return tsconfig?.data?.compilerOptions?.paths || {};
 }
 
-// we treat every package (everything starting . or ..) as external
-// except for typescript aliases we led esbuild to resolve them,
-// but if the alias is dynamic import we treat it as external as well.
+const NON_NODE_MODULE_RE = /^[A-Z]:[/\\]|^\.{0,2}\/|^\.{1,2}$/
 
 function createExternalsPlugin(configPath: string): Plugin {
   const resolvedPaths = tsconfigResolvePaths(configPath);
@@ -24,15 +22,20 @@ function createExternalsPlugin(configPath: string): Plugin {
   return {
     name: "external-packages",
     setup: (build) => {
-      const filter = /^[^.\/]|^\.[^.\/]|^\.\.[^\/]/;
-      build.onResolve({ filter }, ({ path, kind }) => {
+      build.onResolve({filter: /.*/ }, ({ path, kind }) => {
         if (match(path, resolvePatterns)) {
           if (kind === "dynamic-import") {
             return { path, external: true };
           }
           return;
         }
-        return { path, external: true };
+
+        if (!NON_NODE_MODULE_RE.test(path)) {
+          return {
+            path: path,
+            external: true,
+          }
+        }
       });
     },
   };

--- a/packages/core/src/utils.test.ts
+++ b/packages/core/src/utils.test.ts
@@ -1,5 +1,5 @@
-import { describe, expect, it } from "vitest";
-import { generateTypeName, isDefined, removeChildPaths } from "./utils";
+import { describe, expect, it, vitest } from "vitest";
+import { generateTypeName, isDefined, posixToNativePath, removeChildPaths } from "./utils";
 
 describe("generateTypeName", () => {
   it("should return same as collection name", () => {
@@ -78,5 +78,28 @@ describe("removeChildPaths", () => {
     const paths = ["a", "b", "a"];
     const filtered = removeChildPaths(paths);
     expect(filtered).toEqual(["a", "b"]);
+  });
+});
+
+describe("posixToNativePath", () => {
+
+  vitest.mock("node:path", async (importOriginal) => {
+    const origin = await importOriginal<typeof import("node:path")>();
+    return {
+      default: {
+        ...origin,
+        sep: origin.win32.sep,
+      },
+    };
+  });
+
+  it("should replace / with \\", () => {
+    const pathName = posixToNativePath("a/b/c");
+    expect(pathName).toBe("a\\b\\c");
+  });
+
+  it("should not replace \\ with \\", () => {
+    const pathName = posixToNativePath("a\\b\\c");
+    expect(pathName).toBe("a\\b\\c");
   });
 });

--- a/packages/core/src/utils.ts
+++ b/packages/core/src/utils.ts
@@ -1,6 +1,7 @@
 import camelcase from "camelcase";
 import pluralize from "pluralize";
 import { CollectionFile } from "./types";
+import path from "node:path";
 
 export function generateTypeName(name: string) {
   const singularName = pluralize.singular(name);
@@ -28,4 +29,11 @@ export function removeChildPaths(paths: Array<string>) {
       }),
     ),
   );
+}
+
+export function posixToNativePath(pathName: string) {
+  if (path.sep !== path.posix.sep) {
+    return pathName.replaceAll(path.posix.sep, path.sep);
+  }
+  return pathName;
 }

--- a/packages/core/src/watcher.test.ts
+++ b/packages/core/src/watcher.test.ts
@@ -34,7 +34,13 @@ vi.mock("@parcel/watcher", async (importOriginal) => {
   };
 });
 
-describe("watcher", () => {
+describe(
+  "watcher",
+  {
+    // the watcher tests are flaky on windows, so we retry them
+    retry: 3,
+  },
+  () => {
   const events: Array<string> = [];
 
   async function syncFn(modification: Modification, path: string) {

--- a/packages/core/src/watcher.test.ts
+++ b/packages/core/src/watcher.test.ts
@@ -34,11 +34,13 @@ vi.mock("@parcel/watcher", async (importOriginal) => {
   };
 });
 
+const WAIT_UNTIL_TIMEOUT = 5000;
+
 describe(
   "watcher",
   {
     // the watcher tests are flaky on windows, so we retry them
-    retry: 3,
+    retry: 5,
   },
   () => {
   const events: Array<string> = [];
@@ -243,7 +245,7 @@ describe(
 
     await fs.writeFile(path.join(tmpdir, "foo"), "foo");
 
-    await vi.waitUntil(() => findEvent("create", tmpdir, "foo"));
+    await vi.waitUntil(() => findEvent("create", tmpdir, "foo"), WAIT_UNTIL_TIMEOUT);
 
     expect(findEvent("create", tmpdir, "foo")).toBeTruthy();
   });
@@ -260,10 +262,10 @@ describe(
     );
 
     await fs.writeFile(path.join(tmpdir, "foo"), "foo", "utf-8");
-    await vi.waitUntil(() => findEvent("create", tmpdir, "foo"), 2000);
+    await vi.waitUntil(() => findEvent("create", tmpdir, "foo"), WAIT_UNTIL_TIMEOUT);
 
     await fs.writeFile(path.join(tmpdir, "foo"), "bar", "utf-8");
-    await vi.waitUntil(() => findEvent("update", tmpdir, "foo"), 2000);
+    await vi.waitUntil(() => findEvent("update", tmpdir, "foo"), WAIT_UNTIL_TIMEOUT);
 
     expect(findEvent("update", tmpdir, "foo")).toBeTruthy();
   });
@@ -283,7 +285,7 @@ describe(
 
     await fs.rm(path.join(tmpdir, "foo"));
 
-    await vi.waitUntil(() => findEvent("delete", tmpdir, "foo"));
+    await vi.waitUntil(() => findEvent("delete", tmpdir, "foo"), WAIT_UNTIL_TIMEOUT);
 
     expect(findEvent("delete", tmpdir, "foo")).toBeTruthy();
   });
@@ -310,6 +312,7 @@ describe(
       await vi.waitUntil(
         () =>
           findEvent("create", one, "foo") && findEvent("create", two, "bar"),
+        WAIT_UNTIL_TIMEOUT
       );
 
       expect(findEvent("create", one, "foo")).toBeTruthy();
@@ -339,6 +342,7 @@ describe(
       await vi.waitUntil(
         () =>
           findEvent("create", foo, "baz") && findEvent("create", bar, "qux"),
+        WAIT_UNTIL_TIMEOUT
       );
 
       expect(findEvent("create", foo, "baz")).toBeTruthy();
@@ -364,7 +368,7 @@ describe(
       syncFn,
     );
 
-    await vi.waitUntil(() => localEvents.length > 0);
+    await vi.waitUntil(() => localEvents.length > 0, WAIT_UNTIL_TIMEOUT);
 
     expect(localEvents[0]).toBe(`subscribe error:${tmpdir}`);
   });
@@ -405,7 +409,7 @@ describe(
 
     await fs.writeFile(path.join(tmpdir, "baz"), "baz");
 
-    await vi.waitUntil(() => findEvent("create", tmpdir, "baz"));
+    await vi.waitUntil(() => findEvent("create", tmpdir, "baz"), WAIT_UNTIL_TIMEOUT);
 
     expect(findEvent("create", tmpdir, "baz")).toBeTruthy();
   });
@@ -428,7 +432,7 @@ describe(
 
     await watcher?.unsubscribe();
 
-    await vi.waitUntil(() => localEvents.length > 0);
+    await vi.waitUntil(() => localEvents.length > 0, WAIT_UNTIL_TIMEOUT);
     expect(localEvents[0]).toBe(`unsubscribed:${tmpdir}`);
   });
 });

--- a/packages/core/src/watcher.test.ts
+++ b/packages/core/src/watcher.test.ts
@@ -34,13 +34,14 @@ vi.mock("@parcel/watcher", async (importOriginal) => {
   };
 });
 
-const WAIT_UNTIL_TIMEOUT = 5000;
+const WAIT_UNTIL_TIMEOUT = 2000;
 
 describe(
   "watcher",
   {
     // the watcher tests are flaky on windows, so we retry them
-    retry: 5,
+    retry: 3,
+    skip: process.platform === "win32",
   },
   () => {
   const events: Array<string> = [];

--- a/packages/core/src/writer.test.ts
+++ b/packages/core/src/writer.test.ts
@@ -131,81 +131,29 @@ describe("writer", () => {
     expect(existsSync(path.join(tmpdir, "index.d.ts"))).toBe(false);
   });
 
-  describe("os specific", () => {
-    tmpdirTest(
-      "should write import with / instead of \\ on windows",
-      async ({ tmpdir }) => {
-        vitest.mock("node:path", async (importOriginal) => {
-          const origin = await importOriginal<typeof import("node:path")>();
-          return {
-            default: {
-              ...origin,
-              sep: origin.win32.sep,
-              relative: origin.win32.relative,
-            },
-          };
-        });
+  tmpdirTest("should write import with /", async ({ tmpdir }) => {
+      const collections = [
+        {
+          name: "test",
+          typeName: "Test",
+        },
+      ];
 
-        const collections = [
-          {
-            name: "test",
-            typeName: "Test",
-          },
-        ];
+      const writer = await createWriter(tmpdir);
+      await writer.createTypeDefinitionFile({
+        collections,
+        path: path.join(tmpdir, "sub", "config.ts"),
+        generateTypes: true,
+      });
 
-        const writer = await createWriter(tmpdir);
-        await writer.createTypeDefinitionFile({
-          collections,
-          path: path.join(tmpdir, "sub", "config.ts"),
-          generateTypes: true,
-        });
+      const content = await fs.readFile(
+        path.join(tmpdir, "index.d.ts"),
+        "utf-8",
+      );
+      expect(content).toContain(
+        'import configuration from "./sub/config.ts";',
+      );
+    },
+  );
 
-        const content = await fs.readFile(
-          path.join(tmpdir, "index.d.ts"),
-          "utf-8",
-        );
-        expect(content).toContain(
-          'import configuration from "./sub/config.ts";',
-        );
-      },
-    );
-
-    tmpdirTest(
-      "should write import with / on posix based systems",
-      async ({ tmpdir }) => {
-        vitest.mock("node:path", async (importOriginal) => {
-          const origin = await importOriginal<typeof import("node:path")>();
-          return {
-            default: {
-              ...origin,
-              sep: origin.posix.sep,
-              relative: origin.posix.relative,
-            },
-          };
-        });
-
-        const collections = [
-          {
-            name: "test",
-            typeName: "Test",
-          },
-        ];
-
-        const writer = await createWriter(tmpdir);
-        await writer.createTypeDefinitionFile({
-          collections,
-          path: path.join(tmpdir, "sub", "config.ts"),
-          generateTypes: true,
-        });
-
-        const content = await fs.readFile(
-          path.join(tmpdir, "index.d.ts"),
-          "utf-8",
-        );
-        expect(content).toContain(
-          'import configuration from "./sub/config.ts";',
-        );
-      },
-    );
-  });
 });

--- a/turbo.json
+++ b/turbo.json
@@ -14,7 +14,9 @@
       "dependsOn": ["^build"]
     },
     "test": {
-      "dependsOn": ["^build"]
+      "dependsOn": ["^build"],
+      "env": ["CC_OS"]
+
     },
     "typecheck": {
       "dependsOn": ["^build", "^test"]

